day1

ruseel opened the pull request:
https://adventofcode.com/2018/day/1#part2

ruseel committed:
day1 #1

ruseel committed:
destructuring, cycle, cons

kimchhickey commented on clojure/src/aoc2018_1.clj:
```diff
@@ -1,12 +1,45 @@
 ;; 예) +10 -2 -5 +1 이 입력일 경우 4를 출력
 
+(def vs
+  (-> (io/resource "day1.txt")
+      slurp
+      (clojure.string/split #"\n+")
+      (->> (map #(Integer/parseInt %)))))
```
-> 과 ->> 을 함께 사용하는게 좋지 않은 패턴이라고 합니다. 
(Don't mix -> and ->>)
https://stuartsierra.com/2018/07/06/threading-with-style

->와 ->>는 각각 설계의 의도가 있음. map vs collection
-> : assoc, update ...
->> : map, filter, reduce

kimchhickey replied:
`as->`도 있지만, 안씀. 왜냐하면 데이터 구조가 변경되는 로직일 확률이 높음.
let으로 따로 받는 것을 고려하기.

kimchhickey replied:
- 함수형 프로그래밍에서 data-first, data-last에 대한 관점
https://stuartsierra.com/2018/07/06/threading-with-style

ruseel replied:
* 함수형 프로그래밍에서 data-first, data-last에 대한 관점 (링크 수정 본을 기록해둡니다.) 
https://www.javierchavarri.com/data-first-and-data-last-a-comparison/

kimchhickey commented on clojure/src/aoc2018_1.clj:
```diff
@@ -1,12 +1,45 @@
+      (clojure.string/split #"\n+")
+      (->> (map #(Integer/parseInt %)))))
+
+(reduce + vs)
```
`(apply + vs)`

`(+) => 0`
`(+ 1) => 1`

kimchhickey commented on clojure/src/aoc2018_1.clj:
```diff
@@ -1,12 +1,45 @@
 ;; 예) +3, +3, +4, -2, -4 는 10이 처음으로 두번 나오는 숫자임.
 ;; 0 -> 3 (+3) -> 6 (+3) -> 10(+4) -> 8(-2) -> 4(-4) -> 7(+3) -> 10(+3) -> ...
 
+(comment
+
+  (def vs [+1 -2 +3 +1])
+  (def vs [+1 -1])
+  (def vs [+3, +3, +4, -2, -4])
+  (def vs [-6, +3, +8, +5, -6])
+
+)
```
rich comment block 굿입니다!

https://betweentwoparens.com/blog/rich-comment-blocks/

파일 최하단에 한개만 만드는 것이 컨벤션

kimchhickey commented on clojure/src/aoc2018_1.clj:
```diff
@@ -1,12 +1,45 @@
+)
+
+;; first, next 를 사용한 버전
+(loop [seen #{}
+       vs (concat [0] (reductions + (flatten (repeat vs))))]
+  (let [v (first vs)]
+    (if (seen v)
+      v
+      (recur (conj seen v) (next vs)))))
```
`loop-recur` => `reduce`로 바꿔보기

`reduced`

kimchhickey replied:
`goto <-> for-loop <-> reduce <-> map/filter`

kimchhickey replied:
`seen` 👍 


`seen?` 👎 
- 클로저에서 ?는 boolean 리턴할 때만 사용함.

kimchhickey replied:
`reductions` 👍

kimchhickey commented on clojure/src/aoc2018_1.clj:
```diff
@@ -1,12 +1,45 @@
+;;   destructuring, cycle 은 확실히 사용하는 것이 더 좋은 것이다 싶은데요.
+;;   cons 는 좀 애매하다 싶습니다.
+(loop [seen #{}
+       [f & rest] (cons 0 (reductions + (cycle vs)))]
```
`cycle` 👍

kimchhickey replied:
`destructuring` 할수 있는 곳은 한다. (거의 대부분의 경우 가독성이 올라감)

kimchhickey replied:
cons 혹은 seen에 초기값으로 넣기

kimchhickey replied:
`f`라는 이름은 좋을까요? (현재 first-item의 의미임)

https://guide.clojure.style/#idiomatic-names

kimchhickey replied:
rest는 core 함수를 shadowing 할 수 있으므로 변경

kimchhickey replied:
reduce로 추출 후에, find-first-duplicate와 같은 util 함수로 분리해보기

아래 두 개 분리
- 시퀀스 정의
- 비즈니스 로직 (두번 중복되는것 찾기)

ruseel replied:
f & rest 에서 v & more 로 바꿔두었습니다.

ruseel committed:
review

- thread-first, thread-last 섞는 것 없이 좀 더 의도가 잘 드러나게
- loop recur 대신 reduce + reduced 로 구현
- (reduce + ,,,) 를 (apply + ,,,) 로 변경

ruseel committed:
cycled-freq

ruseel committed:
transducer 'dups'
